File: api-cpp/src/Consumer.cpp
#include "rgma/Consumer.h"
#include "rgma/Properties.h"
#include "rgma/ServletConnection.h"

#include <sstream>

namespace glite {
namespace rgma {

int Consumer::create() throw(RGMAPermanentException, RGMATemporaryException) {
    m_connection.addParameter("select", m_query);
    if (m_queryType == QueryType::C || m_queryTypeWithInterval == QueryTypeWithInterval::C) {
        m_connection.addParameter("queryType", std::string("continuous"));
    } else if (m_queryType == QueryType::H || m_queryTypeWithInterval == QueryTypeWithInterval::H) {
        m_connection.addParameter("queryType", std::string("history"));
    } else if (m_queryType == QueryType::L || m_queryTypeWithInterval == QueryTypeWithInterval::L) {
        m_connection.addParameter("queryType", std::string("latest"));
    } else if (m_queryType == QueryType::S) {
        m_connection.addParameter("queryType", std::string("static"));
    }
    if (!(m_queryInterval == TimeInterval::NONE)) {
        m_connection.addParameter("timeIntervalSec", m_queryInterval.getValueAs(TimeUnit::SECONDS));
    }
    if (!(m_timeout == TimeInterval::NONE)) {
        m_connection.addParameter("timeoutSec", m_timeout.getValueAs(TimeUnit::SECONDS));
    }
    std::vector<ResourceEndpoint>::const_iterator rit = m_endpoints.begin();
    while (rit != m_endpoints.end()) {
        std::ostringstream buff;
        buff << rit->getResourceId() << " " << rit->getUrlString();
        m_connection.addParameter("producerConnections", buff.str());
        ++rit;
    }
    TupleSet result;
    m_connection.connect("createConsumer", result);
    TupleSet::const_iterator it = result.begin();
    return it->getInt(0);
}

void Consumer::checkProducerList() const throw (RGMAPermanentException) {
    if (m_endpoints.size() == 0) {
        throw RGMAPermanentException("Consumer constructor with list of producer endpoints given an empty list");
    }
}

Consumer::Consumer(const std::string & query, const QueryType & queryType, const TimeInterval & timeout, std::vector<
        ResourceEndpoint> producers) throw (RGMAPermanentException, RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(queryType), m_timeout(timeout), m_queryTypeWithInterval(
            QueryTypeWithInterval::NONE), m_endpoints(producers), m_queryInterval(TimeInterval::NONE), m_eof(false) {
    checkProducerList();
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryType & queryType, const TimeInterval & timeout)
        throw (RGMAPermanentException, RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(queryType), m_timeout(timeout), m_queryTypeWithInterval(
            QueryTypeWithInterval::NONE), m_endpoints(), m_queryInterval(TimeInterval::NONE), m_eof(false) {
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryType & queryType, std::vector<ResourceEndpoint> producers)
        throw (RGMAPermanentException, RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(queryType), m_timeout(TimeInterval::NONE),
            m_queryTypeWithInterval(QueryTypeWithInterval::NONE), m_endpoints(producers), m_queryInterval(
                    TimeInterval::NONE), m_eof(false) {
    checkProducerList();
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryType & queryType) throw (RGMAPermanentException,
        RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(queryType), m_timeout(TimeInterval::NONE),
            m_queryTypeWithInterval(QueryTypeWithInterval::NONE), m_endpoints(), m_queryInterval(TimeInterval::NONE),
            m_eof(false) {
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryTypeWithInterval & queryType,
        const TimeInterval & queryInterval, const TimeInterval & timeout, std::vector<ResourceEndpoint> producers)
        throw (RGMAPermanentException, RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(QueryType::NONE), m_timeout(timeout), m_queryTypeWithInterval(
            queryType), m_endpoints(producers), m_queryInterval(queryInterval), m_eof(false) {
    checkProducerList();
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryTypeWithInterval & queryType,
        const TimeInterval & queryInterval, const TimeInterval & timeout) throw (RGMAPermanentException,
        RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(QueryType::NONE), m_timeout(timeout), m_queryTypeWithInterval(
            queryType), m_endpoints(), m_queryInterval(queryInterval), m_eof(false) {
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryTypeWithInterval & queryType,
        const TimeInterval & queryInterval, std::vector<ResourceEndpoint> producers) throw (RGMAPermanentException,
        RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(QueryType::NONE), m_timeout(TimeInterval::NONE),
            m_queryTypeWithInterval(queryType), m_endpoints(producers), m_queryInterval(queryInterval), m_eof(false) {
    checkProducerList();
    m_endPoint.setResourceId(create());
}

Consumer::Consumer(const std::string & query, const QueryTypeWithInterval & queryType,
        const TimeInterval & queryInterval) throw (RGMAPermanentException, RGMATemporaryException) :
    Resource("Consumer"), m_query(query), m_queryType(QueryType::NONE), m_timeout(TimeInterval::NONE),
            m_queryTypeWithInterval(queryType), m_endpoints(), m_queryInterval(queryInterval), m_eof(false) {
    m_endPoint.setResourceId(create());
}

Consumer::~Consumer() {
}

void Consumer::abort() throw(RGMATemporaryException, RGMAPermanentException) {
    try {
        doAbort();
    } catch (UnknownResourceException e) {
        try {
            restore();
            doAbort();
            return;
        } catch (UnknownResourceException e1) {
            throw RGMATemporaryException(e1.getMessage());
        }
    }
}

void Consumer::doAbort() throw(RGMAPermanentException, RGMATemporaryException, UnknownResourceException) {
    clearServletConnection();
    TupleSet result;
    m_connection.connect("abort", result);
    checkOK(result);
}

bool Consumer::hasAborted() throw(RGMATemporaryException, RGMAPermanentException) {
    try {
        return doHasAborted();
    } catch (UnknownResourceException e) {
        try {
            restore();
            return doHasAborted();
        } catch (UnknownResourceException e1) {
            throw RGMATemporaryException(e1.getMessage());
        }
    }
}

bool Consumer::doHasAborted() throw (RGMAPermanentException, RGMATemporaryException, UnknownResourceException) {
    clearServletConnection();
    TupleSet result;
    m_connection.connect("hasAborted", result);
    TupleSet::const_iterator it = result.begin();
    return it->getBool(0);
}

void Consumer::pop(int maxCount, TupleSet& results) throw(RGMATemporaryException, RGMAPermanentException) {
    try {
        doPop(maxCount, results);
    } catch (UnknownResourceException e) {
        try {
            restore();
            doPop(maxCount, results);
            results.appendWarning("The query was restarted - many duplicates may be returned.");
        } catch (UnknownResourceException e1) {
            throw RGMATemporaryException(e1.getMessage());
        }
    }
}

void Consumer::doPop(int maxCount, TupleSet& results) throw(RGMAPermanentException, RGMATemporaryException,
        UnknownResourceException) {
    clearServletConnection();
    m_connection.addParameter("maxCount", maxCount);
    m_connection.connect("pop", results);
    if (m_eof) {
        results.appendWarning("You have called pop again after end of results returned.");
    }
    m_eof = results.isEndOfResults();
}

void Consumer::restore() throw (RGMAPermanentException, RGMATemporaryException, UnknownResourceException) {
    Consumer c(m_query, m_queryType, m_timeout, m_queryTypeWithInterval, m_endpoints, m_queryInterval);
    m_endPoint.setResourceId(c.create());
}

Consumer::Consumer(const std::string & query, const QueryType & queryType, const TimeInterval & timeout,
        const QueryTypeWithInterval & queryTypeWithInterval, const std::vector<ResourceEndpoint> & endpoints,
        const TimeInterval & queryInterval) :
    Resource("Consumer"), m_query(query), m_queryType(queryType), m_timeout(timeout), m_queryTypeWithInterval(
            queryTypeWithInterval), m_endpoints(endpoints), m_queryInterval(queryInterval) {
}

}
}

File: trunk/api-java/src/org/glite/rgma/Consumer.java
/*
 * Copyright (c) Members of the EGEE Collaboration. 2004.
 * See http://eu-egee.org/partners/ for details on the copyright holders.
 * For license conditions see the license file or http://eu-egee.org/license.html
 */
package org.glite.rgma;

import java.util.ArrayList;
import java.util.List;

/**
 * Retrieve data from one or more producers by executing a specified SQL query.
 */
public class Consumer extends Resource {

	private String m_query;
	private QueryType m_queryType;
	private TimeInterval m_timeout;
	private QueryTypeWithInterval m_queryTypeWithInterval;
	private List<ResourceEndpoint> m_endpoints;
	private TimeInterval m_queryInterval;
	private boolean m_eof;

	private int create() throws RGMAPermanentException, RGMATemporaryException {
		m_servletName = "ConsumerServlet";
		ServletConnection consumerServlet = new ServletConnection(m_servletName);
		consumerServlet.addParameter("select", m_query);
		if (m_queryType == QueryType.C || m_queryTypeWithInterval == QueryTypeWithInterval.C) {
			consumerServlet.addParameter("queryType", "continuous");
		} else if (m_queryType == QueryType.H || m_queryTypeWithInterval == QueryTypeWithInterval.H) {
			consumerServlet.addParameter("queryType", "history");
		} else if (m_queryType == QueryType.L || m_queryTypeWithInterval == QueryTypeWithInterval.L) {
			consumerServlet.addParameter("queryType", "latest");
		} else if (m_queryType == QueryType.S) {
			consumerServlet.addParameter("queryType", "static");
		}
		if (m_queryInterval != null) {
			consumerServlet.addParameter("timeIntervalSec", m_queryInterval.getValueAs(TimeUnit.SECONDS));
		}
		if (m_timeout != null) {
			consumerServlet.addParameter("timeoutSec", m_timeout.getValueAs(TimeUnit.SECONDS));
		}
		if (m_endpoints != null) {
			for (ResourceEndpoint endpoint : m_endpoints) {
				String endpointString = endpoint.getResourceId() + " " + endpoint.getUrlString();
				consumerServlet.addParameter("producerConnections", endpointString);
			}
		}
		Tuple tuple = consumerServlet.sendNonResourceCommand("createConsumer").getData().get(0);
		return tuple.getInt(0);
	}

	private Consumer(String query, QueryType queryType, TimeInterval timeout, QueryTypeWithInterval queryTypeWithInterval, List<ResourceEndpoint> endpoints,
			TimeInterval queryInterval) {
		m_query = query;
		m_queryType = queryType;
		m_timeout = timeout;
		m_queryTypeWithInterval = queryTypeWithInterval;
		m_endpoints = endpoints;
		m_queryInterval = queryInterval;
	}

	/**
	 * Creates a consumer with the specified query type and where a timeout and list of producers may also be specified.
	 * For a continuous query, only tuples published after the consumer has been created will be returned. All tuples
	 * are included for other query types. The query will terminate after the specified timeout and will make use of the
	 * specified list of producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param timeout
	 *            time interval after which the query will be aborted.
	 * @param producers
	 *            list of producers to contact.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryType queryType, TimeInterval timeout, List<ResourceEndpoint> producers) throws RGMAPermanentException,
			RGMATemporaryException {
		m_query = query;
		m_queryType = queryType;
		m_timeout = timeout;
		m_endpoints = new ArrayList<ResourceEndpoint>(producers);
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and where a timeout may also be specified. For a continuous
	 * query, only tuples published after the consumer has been created will be returned. All tuples are included for
	 * other query types. The query will terminate after the specified timeout and the mediator will be used to find
	 * suitable producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param timeout
	 *            time interval after which the query will be aborted. May be null to have no timeout.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryType queryType, TimeInterval timeout) throws RGMAPermanentException, RGMATemporaryException {
		m_query = query;
		m_queryType = queryType;
		m_timeout = timeout;
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and where a list of producers may also be specified. For a
	 * continuous query, only tuples published after the consumer has been created will be returned. All tuples are
	 * included for other query types. The query will make use of the specified list of producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param producers
	 *            list of producers to contact.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryType queryType, List<ResourceEndpoint> producers) throws RGMAPermanentException, RGMATemporaryException {
		m_query = query;
		m_queryType = queryType;
		m_endpoints = new ArrayList<ResourceEndpoint>(producers);
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type. For a continuous query, only tuples published after the
	 * consumer has been created will be returned. All tuples are included for other query types. The mediator will be
	 * used to find suitable producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryType queryType) throws RGMAPermanentException, RGMATemporaryException {
		m_query = query;
		m_queryType = queryType;
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and query interval and where a timeout and list of producers may
	 * also be specified. The query will terminate after the specified timeout and will make use of the specified list
	 * of producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param queryInterval
	 *            the time interval is subtracted from the current time to give a time in the past. The query result
	 *            will then use data from tuples published after this time.
	 * @param timeout
	 *            time interval after which the query will be aborted.
	 * @param producers
	 *            list of producers to contact.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryTypeWithInterval queryType, TimeInterval queryInterval, TimeInterval timeout, List<ResourceEndpoint> producers)
			throws RGMAPermanentException, RGMATemporaryException {
		m_query = query;
		m_queryTypeWithInterval = queryType;
		m_queryInterval = queryInterval;
		m_timeout = timeout;
		m_endpoints = new ArrayList<ResourceEndpoint>(producers);
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and query interval and where a timeout may also be specified.
	 * The query will terminate after the specified timeout and the mediator will be used to find suitable producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param queryInterval
	 *            the time interval is subtracted from the current time to give a time in the past. The query result
	 *            will then use data from tuples published after this time.
	 * @param timeout
	 *            time interval after which the query will be aborted.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryTypeWithInterval queryType, TimeInterval queryInterval, TimeInterval timeout) throws RGMAPermanentException,
			RGMATemporaryException {
		m_query = query;
		m_queryTypeWithInterval = queryType;
		m_queryInterval = queryInterval;
		m_timeout = timeout;
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and query interval and where a list of producers may also be
	 * specified. The query will make use of the specified list of producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param queryInterval
	 *            the time interval is subtracted from the current time to give a time in the past. The query result
	 *            will then use data from tuples published after this time.
	 * @param producers
	 *            list of producers to contact.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryTypeWithInterval queryType, TimeInterval queryInterval, List<ResourceEndpoint> producers) throws RGMAPermanentException,
			RGMATemporaryException {
		m_query = query;
		m_queryTypeWithInterval = queryType;
		m_queryInterval = queryInterval;
		m_endpoints = new ArrayList<ResourceEndpoint>(producers);
		m_resourceId = create();
	}

	/**
	 * Creates a consumer with the specified query type and query interval. The mediator will be used to find suitable
	 * producers.
	 * 
	 * @param query
	 *            a SQL select statement
	 * @param queryType
	 *            the type of the query
	 * @param queryInterval
	 *            the time interval is subtracted from the current time to give a time in the past. The query result
	 *            will then use data from tuples published after this time.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public Consumer(String query, QueryTypeWithInterval queryType, TimeInterval queryInterval) throws RGMAPermanentException, RGMATemporaryException {
		m_query = query;
		m_queryTypeWithInterval = queryType;
		m_queryInterval = queryInterval;
		m_resourceId = create();
	}

	/**
	 * Aborts the query. Pop can still be called to retrieve existing data.
	 * 
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 * @see #hasAborted
	 */
	public void abort() throws RGMAPermanentException, RGMATemporaryException {
		try {
			doAbort();
		} catch (UnknownResourceException e) {
			try {
				restore();
				doAbort();
				return;
			} catch (UnknownResourceException e1) {
				throw new RGMATemporaryException(e1.getMessage());
			}
		}
	}

	private void doAbort() throws RGMAPermanentException, RGMATemporaryException, UnknownResourceException {
		ServletConnection connection = getNewConnection();
		checkOK(connection.sendCommand("abort"));
	}

	/**
	 * Determines if the query has aborted.
	 * 
	 * @return <code>true</code> if the query was aborted either by a user call or a timeout event
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 * @see #abort
	 */
	public boolean hasAborted() throws RGMAPermanentException, RGMATemporaryException {
		try {
			return doHasAborted();
		} catch (UnknownResourceException e) {
			try {
				restore();
				return doHasAborted();
			} catch (UnknownResourceException e1) {
				throw new RGMATemporaryException(e1.getMessage());
			}
		}
	}

	private boolean doHasAborted() throws RGMAPermanentException, RGMATemporaryException, UnknownResourceException {
		ServletConnection connection = getNewConnection();
		Tuple tuple = connection.sendCommand("hasAborted").getData().get(0);
		return tuple.getBoolean(0);
	}

	/**
	 * Retrieves tuples from the result of the query.
	 * 
	 * @param maxCount
	 *            the maximum number of tuples to retrieve
	 * @return a tuple set containing the received tuples. A set with an empty list of tuples is returned if none were
	 *         found.
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public TupleSet pop(int maxCount) throws RGMAPermanentException, RGMATemporaryException {
		try {
			return doPop(maxCount);
		} catch (UnknownResourceException e) {
			try {
				restore();
				TupleSet ts = doPop(maxCount);
				ts.appendWarning("The query was restarted - many duplicates may be returned.");
				return ts;
			} catch (UnknownResourceException e1) {
				throw new RGMATemporaryException(e1.getMessage());
			}
		}
	}

	private TupleSet doPop(int maxCount) throws RGMAPermanentException, RGMATemporaryException, UnknownResourceException {
		ServletConnection connection = getNewConnection();
		connection.addParameter("maxCount", maxCount);
		TupleSet ts = connection.sendCommand("pop");
		if (m_eof) {
			ts.appendWarning("You have called pop again after end of results returned.");
		}
		m_eof = ts.isEndOfResults();
		return ts;
	}

	private void restore() throws RGMAPermanentException, RGMATemporaryException {
		Consumer c = new Consumer(m_query, m_queryType, m_timeout, m_queryTypeWithInterval, m_endpoints, m_queryInterval);
		m_resourceId = c.create();
	}

}
